lib/service/certreloader_test.go

```diff
@@ -0,0 +1,269 @@
+				require.Equal(t, before[0], after[0])
+				require.Equal(t, before[1], after[1])
+			},
+		},
+		{
+			desc: "c1 cert does not change during an ongoing update",
+			certsUpdate: func(t *testing.T, certs *certFiles) {
+				// Update c1 key, and partially update c1 cert.
+				key, crt := newCertKeyPair(t)
+				write(t, certs.c1Key, key)
+				write(t, certs.c1Crt, crt[0:1024])
+			},
+			certsReloadErrorAssert: require.Error,
+			certsAssert: func(t *testing.T, before []tls.Certificate, after []tls.Certificate) {
+				// No cert has been updated.
+				require.Len(t, before, 2)
+				require.Len(t, after, 2)
+				require.Equal(t, before[0], after[0])
+				require.Equal(t, before[1], after[1])
+			},
+		},
+		{
+			desc: "c1 and c2 certs do not change if one of them is corrupted",
+			certsUpdate: func(t *testing.T, certs *certFiles) {
+				// Corrupt c1 cert key.
+				_, err := certs.c1Key.WriteAt([]byte{1, 2, 3, 4, 5, 6, 7, 8}, 0)
+				require.NoError(t, err)
+				err = certs.c1Key.Sync()
+				require.NoError(t, err)
+			},
+			certsReloadErrorAssert: require.Error,
+			certsAssert: func(t *testing.T, before []tls.Certificate, after []tls.Certificate) {
+				// No cert has been updated.
+				require.Len(t, before, 2)
+				require.Len(t, after, 2)
+				require.Equal(t, before[0], after[0])
+				require.Equal(t, before[1], after[1])
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			// Create empty certs and ensure they get cleaned up.
+			certs := newCerts(t)
+			defer certs.cleanup(t)
+
+			// Start cert reloader.
+			// Set the reload interval to 0 so that the reloading goroutine is not spawned.
+			// This gives us more flexibility in the tests, so that we can call loadCertificates
+			// when we want.
+			cfg := CertReloaderConfig{
+				KeyPairs:               certs.keyPairPaths(),
+				KeyPairsReloadInterval: 0,
+			}
+			certReloader := NewCertReloader(cfg)
+			err := certReloader.Run(context.TODO())
+
+			// Check that certificates load correctly in the synchronous (first) attempt.
+			require.NoError(t, err)
+
+			// Save certs before update.
+			before := make([]tls.Certificate, len(certReloader.certificates))
+			copy(before, certReloader.certificates)
+
+			// Perform cert update.
+			tc.certsUpdate(t, &certs)
+
+			// Perform cert reload.
+			err = certReloader.loadCertificates()
+			tc.certsReloadErrorAssert(t, err)
+
+			// Perform certs assert, passing in the certs before & after the update.
+			tc.certsAssert(t, before, certReloader.certificates)
+		})
+	}
+}
+
+// certFiles contains 2 certificate key pairs.
+type certFiles struct {
+	c1Key *os.File
+	c1Crt *os.File
+	c2Key *os.File
+	c2Crt *os.File
+}
+
+// newCerts creates files for 2 certificate key pairs,
+// generates 2 certificate key pairs, and writes them
+// to the files created.
+func newCerts(t *testing.T) certFiles {
+	// Create files where key-pair certs will be written to.
+	c1Key, err := os.CreateTemp("", "cert1_*.key")
```

Comment:
These files are never closed. I would probably close them before removing them.

Additionally, you can make things even easier on yourself by registering the cleanup call(s) here and removing the `cleanup` func completely.

Another improvement you might find worthwhile is to stop dealing with `*os.File`. Instead you could create a root dir for your certs with `t.TempDir` and then just use `os.WriteFile` to write data to them. This way you don't have to mess with truncating and seeking, and the files will get removed automatically when the test completes.

Reply:
Thanks for the suggestion! Done in 467913190b4330f63350cc8ae0969bcec50696e3.